sdk/trace/batch_span_processor_test.go

```diff
@@ -180,6 +181,48 @@ func TestNewBatchSpanProcessorWithOptions(t *testing.T) {
 	}
 }
 
+func TestBatchSpanProcessorForceFlush(t *testing.T) {
+	option := testOption{
+		name: "ForceFlush()",
+		o: []sdktrace.BatchSpanProcessorOption{
+			sdktrace.WithBatchTimeout(10 * time.Second),
+			sdktrace.WithMaxQueueSize(2000),
+			sdktrace.WithMaxExportBatchSize(2000),
+		},
+		wantNumSpans:   205,
+		wantBatchCount: 1,
+		genNumSpans:    205,
+	}
+
+	te := testBatchExporter{}
+	tp := basicProvider(t)
+	ssp := createAndRegisterBatchSP(option, &te)
+	if ssp == nil {
+		t.Fatalf("%s: Error creating new instance of BatchSpanProcessor\n", option.name)
+	}
+	tp.RegisterSpanProcessor(ssp)
+	tr := tp.Tracer("BatchSpanProcessorWithOptions")
+
+	generateSpan(t, true, tr, option)
+
+	// a small sleep allows for all of the generated spans to be added to the queue
+	time.Sleep(1 * time.Millisecond)
```

Comment:
If we change the `parallel` argument to false it would make this sequential (if I understand this correctly), avoid the need to sleep here, and likely make this more resilient.

Reply:
I thought that at first as well, but it doesn't seem to work out that way:

```console
--- FAIL: TestBatchSpanProcessorForceFlush (0.00s)
    sdk/trace/batch_span_processor_test.go:214: number of exported span: got 175, want 205
FAIL
FAIL	go.opentelemetry.io/otel/sdk/trace	0.004s
FAIL
```

The number of exported spans varies fairly widely when not generating spans in parallel.  When the `parallel` argument is `true` and without the sleep the number of exported spans is fairly reliably 204 of 205.

Reply:
Ah, I see. This sounds like a result of our queueing strategy with the enque being asynchronous from the processing of the batch. Can we add an analogous `drainQueue` call to the `ForceFlush` method prior to exporting the spans that would ensure the flush is complete? I remember looking into this a while ago and realized that it was going to be challenging to solve this problem without a significant refactor of the BatchSpanProcessor itself.

I'm fine with opening an issue to track this as a separate task if you think it is out of scope for this PR.

Reply:
Sounds similar to discussion in #719

Reply:
#719 made me want to never think about this topic again. :)

Reply:
> #719 made me want to never think about this topic again. :)

Sounds like a good reason to create an issue and not block this :laughing:

Reply:
I changed the test expectations to validate that `ForceFlush()` is flushing *any* spans and then unregistering the processor and validating that all the spans were sent.  That avoids the sleep and seems like a reasonable test to me.